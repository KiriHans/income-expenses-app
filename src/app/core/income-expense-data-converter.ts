import {
  FirestoreDataConverter,
  QueryDocumentSnapshot,
  SnapshotOptions,
} from '@angular/fire/firestore';
import { IncomeExpense } from './models/income-expenses.model';

export const incomeExpenseConverter: FirestoreDataConverter<IncomeExpense> = {
  toFirestore: (incomeExpense: IncomeExpense) => {
    const { description, amount, type, uid } = incomeExpense;
    return { description, amount, type, uid };
  },
  fromFirestore: (
    snapshot: QueryDocumentSnapshot<IncomeExpense>,
    options: SnapshotOptions | undefined
  ) => {
    const data = snapshot.data(options);
    return new IncomeExpense(data.description, data.amount, data.type, data.uid);
  },
};
