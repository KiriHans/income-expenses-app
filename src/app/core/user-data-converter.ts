import {
  FirestoreDataConverter,
  QueryDocumentSnapshot,
  SnapshotOptions,
} from '@angular/fire/firestore';
import { UserApp } from './models/users.model';

export const userConverter: FirestoreDataConverter<UserApp> = {
  toFirestore: (user: UserApp) => {
    const { displayName, email } = user;
    return { displayName, email };
  },
  fromFirestore: (
    snapshot: QueryDocumentSnapshot<UserApp>,
    options: SnapshotOptions | undefined
  ) => {
    const data = snapshot.data(options);
    return new UserApp(data.id, data.displayName, data.email);
  },
};
