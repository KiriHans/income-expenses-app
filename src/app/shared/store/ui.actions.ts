import { createActionGroup, emptyProps } from '@ngrx/store';

export const UiActions = createActionGroup({
  source: 'UI',
  events: {
    'Load UI': emptyProps(),
    'Stop Load UI': emptyProps(),
  },
});
