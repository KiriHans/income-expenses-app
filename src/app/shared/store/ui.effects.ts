import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { map, concatMap } from 'rxjs/operators';
import { EMPTY } from 'rxjs';
import { UiActions } from './ui.actions';

@Injectable()
export class UiEffects {
  loadUis$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(UiActions.loadUI),
      concatMap(() =>
        /** An EMPTY observable only emits completion. Replace with your own observable API request */
        EMPTY.pipe(map((data) => UiActions.loadUI()))
      )
    );
  });

  constructor(private actions$: Actions) {}
}
