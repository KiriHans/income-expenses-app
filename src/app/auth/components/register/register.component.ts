import { Component, OnInit, Signal } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterLink } from '@angular/router';
import { FormBuilder, FormGroup, ReactiveFormsModule, Validators } from '@angular/forms';
import { Store } from '@ngrx/store';

import { RegisterActions } from '../../../store/actions/auth.actions';
import { authFeature } from 'src/app/store/reducers/auth.reducer';

@Component({
  selector: 'app-register',
  standalone: true,
  imports: [CommonModule, RouterLink, ReactiveFormsModule],
  templateUrl: './register.component.html',
  styles: ['input { height: 100% }'],
})
export class RegisterComponent implements OnInit {
  registerForm!: FormGroup;
  isLoading!: Signal<boolean>;

  constructor(private fb: FormBuilder, private store: Store) {}

  ngOnInit(): void {
    this.registerForm = this.fb.group({
      name: ['', Validators.required],
      email: ['', [Validators.required, Validators.email]],
      password: ['', Validators.required],
    });

    this.isLoading = this.store.selectSignal(authFeature.selectLoading);
  }

  createUser(): void {
    if (this.registerForm.invalid) return;
    const { name, email, password } = this.registerForm.value;

    this.store.dispatch(RegisterActions.registerWithEmail({ name, email, password }));
  }
}
