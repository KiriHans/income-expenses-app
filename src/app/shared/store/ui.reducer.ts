import { createFeature, createReducer, on } from '@ngrx/store';
import { UiActions } from './ui.actions';

export const uiFeatureKey = 'ui';

export interface UIState {
  isLoading: boolean;
}

export const initialState: UIState = {
  isLoading: false,
};

export const uiReducer = createReducer(
  initialState,
  on(UiActions.loadUI, (state): UIState => ({ ...state, isLoading: true })),
  on(UiActions.stopLoadUI, (state): UIState => ({ ...state, isLoading: false }))
);

export const uiFeature = createFeature({
  name: uiFeatureKey,
  reducer: uiReducer,
});
