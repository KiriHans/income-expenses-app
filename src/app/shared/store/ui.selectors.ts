import { createFeatureSelector } from '@ngrx/store';
import * as fromUi from './ui.reducer';

export const selectUiState = createFeatureSelector<fromUi.UIState>(fromUi.uiFeatureKey);
