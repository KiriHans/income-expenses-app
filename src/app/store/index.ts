import { isDevMode } from '@angular/core';
import { ActionReducerMap, MetaReducer } from '@ngrx/store';
import * as ui from '../shared/store/ui.reducer';
import * as auth from './reducers/auth.reducer';
import * as incomeExpense from './reducers/income-expense.reducer';

export const stateFeatureKey = 'state';

export interface AppState {
  [ui.uiFeatureKey]: ui.UIState;
  [auth.authFeatureKey]: auth.State;
  [incomeExpense.incomeExpenseFeatureKey]: incomeExpense.State;
}

export const reducers: ActionReducerMap<AppState> = {
  [ui.uiFeatureKey]: ui.uiFeature.reducer,
  [auth.authFeatureKey]: auth.userReducer,
  [incomeExpense.incomeExpenseFeatureKey]: incomeExpense.reducer,
};

export const metaReducers: MetaReducer<AppState>[] = isDevMode() ? [] : [];
