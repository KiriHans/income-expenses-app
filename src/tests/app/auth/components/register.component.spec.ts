import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Router, provideRouter } from '@angular/router';

import { provideMockStore, MockStore } from '@ngrx/store/testing';
import { RegisterComponent } from 'src/app/auth/components/register/register.component';
import { DumbMockComponent } from '../../mocks/dumb.component.mock';
import { DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';

describe('RegisterComponent', () => {
  let component: RegisterComponent;
  let fixture: ComponentFixture<RegisterComponent>;
  let el: DebugElement;
  let compiled: HTMLElement;

  let store: MockStore;
  const initialState = { user: { loading: false } } as const;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [RegisterComponent],
      providers: [
        provideMockStore({ initialState }),
        provideRouter([{ path: 'auth/login', component: DumbMockComponent }]),
      ],
    });
    fixture = TestBed.createComponent(RegisterComponent);
    component = fixture.componentInstance;
    el = fixture.debugElement;
    compiled = fixture.nativeElement;

    store = TestBed.inject(MockStore);

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should render a form with email and password inputs', () => {
    const sigUpTitle = el.query(By.css('h2'));
    const registerForm = el.query(By.css('form'));
    const [nameInput, emailInput, passwordInput] = el.queryAll(By.css('.form-group input'));

    expect(sigUpTitle.nativeElement.textContent).toBe('Register');
    expect(registerForm.nativeElement.textContent).toBeTruthy();

    expect(nameInput.name).toBe('input');
    expect(nameInput.attributes['type']).toBe('text');
    expect(nameInput.attributes['formControlName']).toBe('name');
    expect(nameInput.attributes['placeholder']).toBe('Name');

    expect(emailInput.name).toBe('input');
    expect(emailInput.attributes['type']).toBe('email');
    expect(emailInput.attributes['formControlName']).toBe('email');
    expect(emailInput.attributes['placeholder']).toBe('Email');

    expect(passwordInput.name).toBe('input');
    expect(passwordInput.attributes['type']).toBe('password');
    expect(passwordInput.attributes['formControlName']).toBe('password');
    expect(passwordInput.attributes['placeholder']).toBe('Password');
  });

  it('should have a button that is disabled when form is invalid', () => {
    const submitButton = el.query(By.css('.form-group button.submit-btn'));

    expect(submitButton.nativeElement.disabled).toBeTruthy();
    expect(submitButton.name).toBe('button');
    expect(submitButton.classes['submit-btn']).toBeTruthy();
    expect(submitButton.nativeElement.textContent.trim()).toBe('Create account');
  });

  it('should have a description', () => {
    const descText = el.query(By.css('.text-block span'));

    expect(descText.nativeElement.textContent.trim()).toBe('Already have a new account?');
  });

  it('should have a link that redirects to the register page', () => {
    const link = el.query(By.css('span~a[routerLink="/auth/login"]'));

    (link.nativeElement as HTMLAnchorElement).click();

    expect(TestBed.inject(Router).url).toEqual('/auth/login');
  });

  it('should enable button when form is valid', () => {
    const submitButton = el.query(By.css('.form-group button.submit-btn'));

    component.registerForm.setValue({
      name: 'test',
      email: 'test@email.com',
      password: '123456',
    });

    fixture.detectChanges();

    expect(submitButton.nativeElement.disabled).toBeFalsy();
  });

  it('should log in user when form is valid and button is pressed', () => {
    const submitButton = el.query(By.css('.form-group button.submit-btn'));
    const registerSpy = jest.spyOn(component, 'createUser');
    const storeSpy = jest.spyOn(store, 'dispatch');

    component.registerForm.setValue({
      name: 'test',
      email: 'test@email.com',
      password: '123456',
    });

    fixture.detectChanges();
    (submitButton.nativeElement as HTMLButtonElement).click();

    expect(registerSpy).toHaveBeenCalled();
    expect(storeSpy).toHaveBeenCalled();
    expect(submitButton.nativeElement.disabled).toBeFalsy();
  });

  it('should show new button with "waiting..." text while user is authenticating', () => {
    let submitButton = el.query(By.css('.form-group button.submit-btn'));

    component.registerForm.setValue({
      name: 'test',
      email: 'test@email.com',
      password: '123456',
    });

    fixture.detectChanges();

    store.setState({ user: { loading: true } });

    store.refreshState();

    fixture.detectChanges();

    submitButton = el.query(By.css('.form-group button.submit-btn'));

    expect(submitButton.nativeElement.textContent.trim()).toBe('Waiting...');
  });
});
