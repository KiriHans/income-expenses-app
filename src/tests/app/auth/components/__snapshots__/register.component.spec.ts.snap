// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="page-body-wrapper full-page-wrapper auth-page"
  >
    <div
      class="content-wrapper d-flex align-items-center auth register-bg-1 theme-one"
    >
      <div
        class="row w-100"
      >
        <div
          class="col-lg-4 mx-auto"
        >
          <h2
            class="text-center mb-4"
          >
            Register
          </h2>
          <div
            class="auto-form-wrapper"
          >
            <form
              class="ng-untouched ng-pristine ng-invalid"
              novalidate=""
            >
              <div
                class="form-group"
              >
                <div
                  class="input-group"
                >
                  <input
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="name"
                    placeholder="Name"
                    type="text"
                  />
                  <div
                    class="input-group-append"
                  >
                    <span
                      class="input-group-text"
                    >
                      
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="form-group"
              >
                <div
                  class="input-group"
                >
                  <input
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="email"
                    placeholder="Email"
                    type="email"
                  />
                  <div
                    class="input-group-append"
                  >
                    <span
                      class="input-group-text"
                    >
                      
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="form-group"
              >
                <div
                  class="input-group"
                >
                  <input
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="password"
                    placeholder="Password"
                    type="password"
                  />
                  <div
                    class="input-group-append"
                  >
                    <span
                      class="input-group-text"
                    >
                      
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="form-group"
              >
                <button
                  class="btn btn-primary submit-btn btn-block"
                  disabled=""
                >
                   Create account 
                </button>
                
                
              </div>
              <div
                class="text-block text-center my-3"
              >
                <span
                  class="text-small font-weight-semibold"
                >
                  Already have a new account? 
                </span>
                <br />
                <a
                  class="text-black text-small"
                  href="/auth/login"
                  routerlink="/auth/login"
                >
                  Login
                </a>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
